스터디 기록 작성 API

12PJH opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #47 
- 작업 요약: 스터디 기록 작성에 대한 파일 생성 및 작업

## 📄 상세 내용
- [x] view 생성
- [x] serializer 생성
- [x] service 생성
- [x] url 생성
- [x] testcode 생성

## 📸 스크린샷 (선택)
<img width="1342" height="778" alt="스크린샷 2025-09-04 17 03 06" src="https://github.com/user-attachments/assets/1780761f-2700-4db5-9c24-bd7bd8c2553e" />

## 📝 기타 참고 사항
스터디 기록 CRUD 부분 중 "C" 부분 작성 완료하였습니다.

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

12PJH committed:
♻️ refactor : reposittory 폴더 삭제

12PJH committed:
✨ feat : mockdata 및 testcode 작성(스터디 학습 기록 생성)

12PJH committed:
♻️ refactor : views.py => study_notes_views.py 이름 변경

12PJH committed:
♻️ refactor : code_formatting

12PJH committed:
♻️ refactor : 리팩터링 (코드 포멧팅)

12PJH committed:
♻️ refactor : 리팩터링 (url 네임 변경 후 적용)

12PJH committed:
✨ feat: StudyNoteSerializer 기본 필드 정의

1. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- study_group, author, title, content, images, attachments, ai_summary 필드 포함
- read_only_fields 설정으로 id, created_at, updated_at 보호
- images(ListField 사용 / 여러 이미지를 받기위해)
- attachments(DictField 사용 / 파일과 파일 이름을 같이 받기위해)

12PJH committed:
✨ feat: StudyNoteService 생성 및 create_study_note 구현

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- StudyNote 생성 로직 Service로 분리
- 뷰에서 서비스 호출

12PJH committed:
✨ feat: StudyNoteCreateView 구현

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- POST 요청
- Serializer 유효성 검증 후 Service 호출
- 생성된 StudyNote 반환

12PJH committed:
✨ feat: StudyNoteCreate URL 설정 및 config 적용

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- /api/v1/studies/<group_uuid>/notes 엔드포인트 생성
- DRF view, URL conf 등록 및 config 설정 완료

12PJH committed:
♻️ refactor : 리팩터링 (스터디 기록 작성 url 파일 추적 설정 변경)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- init 파일에 urlpatterns 임포트
- config url에 include 부분 수정

12PJH committed:
♻️ refactor : 리팩터링 (urls에 app_name 추가)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- apps.urls에 app_name 추가
- Config에 namespace 추가

12PJH committed:
♻️ refactor : 리팩터링 (오류 나서 롤백)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 추가적으로 수정해볼려다 오류나서 롤백

12PJH committed:
🎨 style    : 코드 포매팅 수정

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 1차적으로 black, isort, mypy는 오류 없이 성공
- coverage 오류 수정 후 다시 테스트 예정

12PJH committed:
✅ test     : 테스트 코드 추가(스터디 그룹 기록 작성)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 기록 작성 시 필수값 누락 시 나오는 에러 확인 코드

12PJH committed:
🎨 style    : 코드 포매팅 수정

12PJH committed:
♻️ refactor : 리팩터링 (테스트 코드)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 실패 테스트 반응 수정

12PJH committed:
🐛 fix : config / urls 충돌 해결

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- urlpatterns 추가 부분 충돌

Meoyoug commented on apps/study_notes/serializers/study_notes_serializers.py:
```diff
@@ -0,0 +1,32 @@
+    images = serializers.ListField(child=serializers.URLField(), write_only=True, required=False)
+    # 여러 첨부파일은 파일 이름도 같이 들어오기 때문에 DictField로 받음
+    attachments = serializers.ListField(
+        child=serializers.DictField(child=serializers.CharField()), write_only=True, required=False
```
여러 첨부파일은.. ListField + FileField의 조합으로 받을 수 있습니다

12PJH replied:
넵 수정하겠습니다.

Meoyoug commented on apps/study_notes/serializers/study_notes_serializers.py:
```diff
@@ -0,0 +1,32 @@
+    """
+
+    # 여러 이미지가 들어올 수 있기 떄문에 ListField로 받음
+    images = serializers.ListField(child=serializers.URLField(), write_only=True, required=False)
```
여기도 마찬가지로 ListField + ImageField를 활용해서 여러 이미지를 받을 수 있어요

Meoyoug commented on apps/study_notes/services/study_notes_services.py:
```diff
@@ -0,0 +1,47 @@
+            ai_summary="AI 요약은 추후 자동 생성됩니다.",
+        )
+
+        # 이미지 저장
```
이부분 s3에 업로드하는걸로 바꿔주세요
core/utils/s3uploader.py 만들어서 사용해주세요

Meoyoug commented on apps/study_notes/views/study_notes_views.py:
```diff
@@ -0,0 +1,44 @@
+        group = get_object_or_404(StudyGroup, uuid=group_uuid)
+        serializer = StudyNoteSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        print(serializer.errors)  # 테스트 결과에 오류가 있을 경우 출력
```
print문은 제거부탁드려요

Meoyoug commented on apps/study_notes/tests/Mockdata.py:
얘는 어디서 쓰이는 앨까요..?

Meoyoug commented on apps/study_notes/tests/testcode.py:
파일 이름 test_테스트목적이나내용.py 로 컨벤션 맞춰주세요 ci 에서 안잡혀요

12PJH committed:
🐛 fix : config / urls 충돌 해결
♻️ refactor : 리팩터링 (이미지, 파일 시리얼라이저 필드 수정)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 이미지 부분 ListField + ImageField 사용
- 파일 부분 ListField + FileField 사용

12PJH committed:
♻️ refactor : 리팩터링 (PR 피드백 수정)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 불필요한 print문 삭제
- test 파일 목적에 맞게 이름 변경

SimJunSik commented on apps/study_notes/services/study_notes_services.py:
```diff
@@ -0,0 +1,47 @@
+            ai_summary="AI 요약은 추후 자동 생성됩니다.",
+        )
+
+        # 이미지 저장
+        if images:
+            for url in images:
+                StudyNoteImage.objects.create(study_note=note, img_url=url)
+
+        # 파일 저장
+        if attachments:
+            for attach in attachments:
+                StudyNoteAttachment.objects.create(
+                    study_note=note, file_url=attach["file_url"], file_name=attach["file_name"]
+                )
```
ORM 코드, 그러니까 DB 호출 코드가 반복문 안에 있다면, 아니면 꼭 반복문이 아니더라도 구조적으로 반복된다면, 호출을 줄일 수 없을지 꼭 한 번 생각해보세요

DB 호출이 리소스적으로 비싸서 이런 부분에서 성능에 악영향을 많이 끼칩니다

12PJH replied:
넵 방법 찾아본 뒤 수정하겠습니다!

SimJunSik commented on apps/study_notes/tests/Mockdata.py:
```diff
@@ -0,0 +1,100 @@
+# 유저 생성
+def create_mock_users(count: int = 10) -> List[User]:
+    users = []
+    for i in range(count):
```
이런 테스트 코드에서도 DB 호출을 줄여보세요
호출을 줄이라는 건, 한 번의 query 로 할 수 없을지 입니다
테스트 코드라도, 이런게 쌓이면 CI 도는데에 시간이 점점 많이 늘어나고 이게 곧 생산성에 영향을 많이 끼치게 될거에요

12PJH committed:
♻️ refactor : 리팩터링 (이미지, 파일 쪽 저장 루트 변경)

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
1. S3쪽으로 업로드 방향 수정

12PJH committed:
Merge remote-tracking branch 'origin/develop' into feature/study_notes_test

12PJH committed:
✅ test : 테스트 코드 추가/변경 (스터디 기록 생성 테스트 코드)

변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- 불필요한 Mockdata 파일 삭제
- 이미지 및 첨부 파일 S3 업로더 적용(테스트용 mock_s3_uploader 생성)
- ORM, managers.py로 이동 및 사용
- 생성 테스트, 필수값 누락 시 실패 테스트 작성

12PJH committed:
Merge remote-tracking branch 'origin/develop' into feature/study_notes_test

Umdoong commented on apps/study_notes/managers.py:
```diff
@@ -0,0 +1,33 @@
+            study_group=study_group,
+            title=title,
+            content=content,
+            ai_summary="AI 요약은 추후 자동 생성됩니다.",
```
ai_summary="AI 요약은 추후 자동 생성됩니다."
이건 지워도 될 것 같아용 ㅋㅋㅋㅋㅋ

Umdoong commented on apps/study_notes/serializers/study_notes_serializers.py:
```diff
@@ -0,0 +1,24 @@
+
+
+class StudyNoteSerializer(serializers.ModelSerializer[StudyNote]):
+    images = serializers.ListField(child=serializers.ImageField(), write_only=True, required=False)
```
images와 attachments가 swagger 문서에서도 모델에 맞게 나올 수 있도록
Nasted Serilizer로 해주시면 좋을 것 같아요

Umdoong reviewed (changes requested):
conflict 났어용 ㅠ
develop을 pull(rebase)해서 최신화한 후 해결해서 다시 push해주세요!

12PJH committed:
✅ test : 테스트 코드 변경 (스터디 기록 작성 API)

1. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
- Nasted Serilizer 수정
- 테스트용 mock_s3_uploader 사용
